### ISSUE TITLE
inconsistent of einsum and torch.mm

### ISSUE THREAD

**Issue by Anonymousjoy** [pytorch/pytorch#27016]
## Issue description
When comparing the outcomes of torch.mm and torch.einsum for matrix multiplication, the results is not consistent. The same result is also produced by numpy. 

## Code example
A minimal example is down here. 
`import torch
import numpy as np
torch.manual_seed(2)
a = torch.randn(4, 4)
b = torch.randn(4, 4)
c_torch_mm = torch.mm(a, b)
c_torch_einsum = torch.einsum('ik,kj -> ij', a, b)
a_ = a.numpy()
b_ = b.numpy()
c_np_matmul = np.matmul(a_, b_)
c_np_einsum = np.einsum('ik, kj -> ij', a_, b_)

print(c_torch_mm.numpy() == c_np_matmul)
print(c_torch_einsum.numpy() == c_np_einsum)
print(c_torch_mm == c_torch_einsum)
print(c_torch_mm.numpy() == c_np_einsum)`

[[ True  True  True  True]
 [ True  True  True  True]
 [ True  True  True  True]
 [ True  True  True  True]]
[[ True  True  True  True]
 [ True  True  True  True]
 [ True  True  True  True]
 [ True  True  True  True]]
tensor([[ True,  True,  True,  True],
        [False,  True, False,  True],
        [ True, False,  True,  True],
        [False,  True,  True,  True]])
[[ True  True  True  True]
 [False  True False  True]
 [ True False  True  True]
 [False  True  True  True]]

While of course we should expect all trues.

## System Info
PyTorch version: 1.2.0
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Red Hat Enterprise Linux release 8.0 (Ootpa)
GCC version: (GCC) 8.2.1 20180905 (Red Hat 8.2.1-3)
CMake version: version 3.11.4

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.1.168
GPU models and configuration: GPU 0: GeForce RTX 2080 Ti
Nvidia driver version: 430.14
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.16.2
[pip] numpydoc==0.8.0
[pip] torch==1.2.0
[conda] _pytorch_select           0.2                       gpu_0  
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.4                      243  
[conda] mkl-service               2.3.0            py37he904b0f_0  
[conda] mkl_fft                   1.0.10           py37ha843d7b_0  
[conda] mkl_random                1.0.2            py37hd81dba3_0  
[conda] pytorch                   1.2.0           cuda100py37h938c94c_0

**Comment by t-vi** [comment]
Please don't compare numerical computation results with `==`, but check with a certain tolerance. Floating point calculation results are only well-defined up to numerical precision.

**Comment by yaroslavvb** [comment]
The error of your floating point computation may be affected by various factors, such as [other programs](https://www.nag.com/content/wandering-precision) running on the machine.

However, the error is still bounded, so whether exact equality testing will succeed depends on the specific values. 
 
First of all, note that two results are off by exactly 1 ulp in 4 places

```
import numpy.testing._private.utils as nputil
nputil.nulp_diff(c_torch_mm.numpy(), c_np_einsum) 
```
gives
```
array([[0., 0., 0., 0.],
       [1., 0., 1., 0.],
       [0., 1., 0., 0.],
       [1., 0., 0., 0.]], dtype=float32)
```

Is this acceptable error for a standard matrix multiplication algorithm? You can use approach from numerical analysis to check, ie, Higham 3.5 "Accuracy and Stability of Numerical Algorithms"

<img width="539" alt="Screenshot 2019-09-28 10 19 22" src="https://user-images.githubusercontent.com/23068/65820064-79da1900-e1d9-11e9-8199-fd58a0a60149.png">
<img width="516" alt="Screenshot 2019-09-28 10 19 28" src="https://user-images.githubusercontent.com/23068/65820067-7e9ecd00-e1d9-11e9-9c46-f0cc7cd05c92.png">

To apply this analysis to your case, 
```
def float32_matmul_error(a, b):
    a = a.type(torch.float64)
    b = b.type(torch.float64)
    n = max(a.shape+b.shape)
    u = np.finfo(np.dtype('float32')).eps
    gamma=(n*u)/(1-n*u)
    error=torch.norm(a.flatten())*torch.norm(b.flatten())*gamma
    return error

scipy.linalg.norm(c_torch_mm.numpy()-c_np_einsum, 'fro')/err  #=> 23.83
```

In other words the discrepancy you are seeing is about 23 times less than worst case error of standard matrix multiplication.

**Comment by Anonymousjoy** [comment]
Thanks you guys so much, it really bothers me a lot.

**Comment by yaroslavvb** [comment]
Here's the utility I personally for equality testing, it's `check_equal` below
```
def pytorch_dtype_to_floating_numpy_dtype(dtype):
    """Converts PyTorch dtype to numpy floating point dtype, defaulting to np.float32 for non-floating point types."""
    if dtype == torch.float64:
        dtype = np.float64
    elif dtype == torch.float32:
        dtype = np.float32
    elif dtype == torch.float16:
        dtype = np.float16
    else:
        dtype = np.float32
    return dtype

def to_numpy(x, dtype: np.dtype = None) -> np.ndarray:
    """
    Convert numeric object to floating point numpy array. If dtype is not specified, use PyTorch default dtype.

    Args:
        x: numeric object
        dtype: numpy dtype, must be floating point

    Returns:
        floating point numpy array
    """

    assert np.issubdtype(dtype, np.floating), "dtype must be real-valued floating point"

    # Convert to normal_form expression from a special form (https://reference.wolfram.com/language/ref/Normal.html)
    if hasattr(x, 'normal_form'):
        x = x.normal_form()

    if type(x) == np.ndarray:
        assert np.issubdtype(x.dtype, np.floating), f"numpy type promotion not implemented for {x.dtype}"

    if type(x) == torch.Tensor:
        dtype = pytorch_dtype_to_floating_numpy_dtype(x.dtype)
        return x.detach().cpu().numpy().astype(dtype)

    # list or tuple, iterate inside to convert PyTorch arrrays
    if type(x) in [list, tuple]:
        x = [to_numpy(r) for r in x]

    # Some Python type, use numpy conversion
    result = np.array(x, dtype=dtype)
    assert np.issubdtype(result.dtype, np.number), f"Provided object ({result}) is not numeric, has type {result.dtype}"
    if dtype is None:
        return result.astype(pytorch_dtype_to_floating_numpy_dtype(torch.get_default_dtype()))
    return result


def check_close(a0, b0, rtol=1e-5, atol=1e-8, label: str= '') -> None:
    """Convenience method for check_equal with tolerances defaulting to typical errors observed in neural network
    ops in float32 precision."""
    return check_equal(a0, b0, rtol=rtol, atol=atol, label=label)


def check_equal(observed, truth, rtol=1e-9, atol=1e-12, label: str= '') -> None:
    """
    Assert fail any entries in two arrays are not close to each to desired tolerance. See np.allclose for meaning of rtol, atol

    """

    # special handling for lists, which could contain
    #if type(observed) == List and type(truth) == List:
    #    for a, b in zip(observed, truth):
    #        check_equal(a, b)

    truth = to_numpy(truth)
    observed = to_numpy(observed)

    assert truth.shape == observed.shape, f"Observed shape {observed.shape}, expected shape {truth.shape}"
    # run np.testing.assert_allclose for extra info on discrepancies
    if not np.allclose(observed, truth, rtol=rtol, atol=atol, equal_nan=True):
        print(f'Numerical testing failed for {label}')
        np.testing.assert_allclose(truth, observed, rtol=rtol, atol=atol, equal_nan=True)

```
